sdk/core/System.ClientModel/src/ModelReaderWriter/UnsafeBufferSequence.Reader.cs

```diff
@@ -0,0 +1,125 @@
+            : base(buffers, count)
+        {
+        }
+    }
+
+    internal class Reader : IDisposable
+    {
+        private readonly BufferSegment[] _buffers;
+        private readonly int _count;
+        private bool _isDisposed;
+        private static readonly object _disposeLock = new object();
+
+        private protected Reader(BufferSegment[] buffers, int count)
+        {
+            _buffers = buffers;
+            _count = count;
+        }
+
+        public long Length
+        {
+            get
+            {
+                if (_isDisposed)
+                {
+                    throw new ObjectDisposedException(nameof(Reader));
+                }
+
+                long result = 0;
+                for (int i = 0; i < _count; i++)
+                {
+                    result += _buffers[i].Written;
+                }
+                return result;
+            }
+        }
+
+        public void CopyTo(Stream stream, CancellationToken cancellation)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(Reader));
+            }
+
+            for (int i = 0; i < _count; i++)
+            {
+                cancellation.ThrowIfCancellationRequested();
+
+                BufferSegment buffer = _buffers[i];
+                stream.Write(buffer.Array, 0, buffer.Written);
+            }
+        }
+
+        public async Task CopyToAsync(Stream stream, CancellationToken cancellation)
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(Reader));
+            }
+
+            for (int i = 0; i < _count; i++)
+            {
+                cancellation.ThrowIfCancellationRequested();
+
+                BufferSegment buffer = _buffers[i];
+                await stream.WriteAsync(buffer.Array, 0, buffer.Written, cancellation).ConfigureAwait(false);
+            }
+        }
+
+        public BinaryData ToBinaryData()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(Reader));
+            }
+
+            long length = Length;
+            if (length > int.MaxValue)
+            {
+                throw new InvalidOperationException($"Length of serialized model is too long.  Value was {length} max is {int.MaxValue}");
+            }
+            using var stream = new MemoryStream((int)length);
+            CopyTo(stream, default);
+            return new BinaryData(stream.GetBuffer().AsMemory(0, (int)stream.Position));
+        }
+
+        public void Dispose()
+        {
+            if (!_isDisposed)
+            {
+                lock (_disposeLock)
+                {
+                    if (!_isDisposed)
+                    {
+                        for (int i = 0; i < _count; i++)
+                        {
+                            ArrayPool<byte>.Shared.Return(_buffers[i].Array);
```

Comment:
would it be beneficial to also set `_buffers[i].Written = 0` to mitigate copy issues during dispose?

Reply:
This isn't a bad idea, it helps minimize the disposing while copying issue

Reply:
we should set _buffers to null or empty. We definitely don't want to hold to the buffers after we return them.

Reply:
I thought about that too, but these are all structs, I thought